command/agent/exec/exec.go

```diff
@@ -0,0 +1,260 @@
+package exec
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"os"
+	"time"
+
+	"github.com/hashicorp/consul-template/child"
+	ctconfig "github.com/hashicorp/consul-template/config"
+	"github.com/hashicorp/consul-template/manager"
+	"github.com/hashicorp/go-hclog"
+
+	"github.com/hashicorp/vault/command/agent/config"
+	"github.com/hashicorp/vault/command/agent/internal/ctmanager"
+	"github.com/hashicorp/vault/helper/useragent"
+	"github.com/hashicorp/vault/sdk/helper/pointerutil"
+)
+
+type processState uint8
+
+const (
+	processStateNotStarted processState = iota
+	processStateRunning
+	processStateRestarting
+	processStateStopped
+)
+
+type ServerConfig struct {
+	Logger      hclog.Logger
+	AgentConfig *config.Config
+
+	Namespace string
+
+	// LogLevel is needed to set the internal Consul Template Runner's log level
+	// to match the log level of Vault Agent. The internal Runner creates it's own
+	// logger and can't be set externally or copied from the Template Server.
+	//
+	// LogWriter is needed to initialize Consul Template's internal logger to use
+	// the same io.Writer that Vault Agent itself is using.
+	LogLevel  hclog.Level
+	LogWriter io.Writer
+}
+
+type Server struct {
+	// config holds the ServerConfig used to create it. It's passed along in other
+	// methods
+	config *ServerConfig
+
+	// runner is the consul-template runner
+	runner *manager.Runner
+
+	// numberOfTemplates is the count of templates determined by consul-template,
+	// we keep the value to ensure all templates have been rendered before
+	// starting the child process
+	// NOTE: each template may have more than one TemplateConfig, so the numbers may not match up
+	numberOfTemplates int
+
+	logger hclog.Logger
+
+	proc      *child.Child
+	procState processState
+
+	// exit channel of the child process
+	procExitCh <-chan int
```

Comment:
Nit: maybe this could be more descriptive if the name of the variable included that it was of the child process, e.g. `childProcExitCh`?

Reply:
Agreed, this would be a better name, same for other procX variables.

Reply:
`childProc` or should we do all the way out -> `childProcess`
eg: `childProcessExitCh`

Reply:
answer in slack:
write it all the way out
(also updated)